src/goto-harness/goto_harness_generator.h

```diff
@@ -0,0 +1,34 @@
+/******************************************************************\
+
+Module: goto_harness_generator
+
+Author: Diffblue Ltd.
+
+\******************************************************************/
+
+#ifndef CPROVER_GOTO_HARNESS_GOTO_HARNESS_GENERATOR_H
+#define CPROVER_GOTO_HARNESS_GOTO_HARNESS_GENERATOR_H
+
+#include <goto-programs/goto_model.h>
+#include <util/cmdline.h>
+#include <util/irep.h>
+
+class goto_harness_generatort
+{
+public:
+  /// Generate a harness according to the set options
+  virtual void
+  generate(goto_modelt &goto_model, const irep_idt &harness_function_name) = 0;
+
+  virtual ~goto_harness_generatort() = default;
+  friend class goto_harness_generator_factoryt;
+
+protected:
+  /// Handle a command line argument. Should throw an exception if the option
+  /// doesn't apply to this generator. Should only set options rather than
+  /// immediately performing work
+  virtual void
+  handle_option(const std::string &option, const std::string &value) = 0;
```

Comment:
Would it make sense to add a method `finalise_option_handling()` that is invoked after all the `handle_option()` calls? I'm thinking of the case when there are required options or dependencies between options (such as if `--a` is present `--b` needs to be present as well). Then those checks could be done there instead of in `generate()`.

Reply:
It might make sense to do that yes, we'll have to think about it